File: native/android/voice/sakha/SakhaTypes.kt
/**
 * Sakha Voice — Shared Types
 *
 * Data classes, enums, and configuration for the Sakha Voice mode pipeline.
 * Sakha (सखा) is the friend/charioteer voice persona defined in
 * sakha.voice.openai.md — distinct from the generic KIAAN voice manager.
 *
 * Voice mode = mic → STT → backend Sakha SSE → pause-aware Sarvam TTS playback.
 */

package com.mindvibe.kiaan.voice.sakha

import java.util.Locale

// ============================================================================
// Engine Selection
// ============================================================================

/**
 * Which Sakha engine should the backend route to. Mirrors the <engine> tag
 * from the Sakha system prompt.
 */
enum class SakhaEngine(val wire: String) {
    GUIDANCE("GUIDANCE"),    // Wisdom-forward — verses front and centre
    FRIEND("FRIEND"),        // Empathetic, witnessing
    ASSISTANT("ASSISTANT"),  // Tool use, navigation
    VOICE_GUIDE("VOICE_GUIDE"); // Acknowledgment when handing off to a tool

    companion object {
        fun fromWire(value: String?): SakhaEngine = when (value?.uppercase()) {
            "GUIDANCE" -> GUIDANCE
            "ASSISTANT" -> ASSISTANT
            "VOICE_GUIDE" -> VOICE_GUIDE
            else -> FRIEND
        }
    }
}

/**
 * Mood label that drives tone selection. The 10 moods from the Sakha persona.
 */
enum class SakhaMood(val wire: String) {
    ANXIOUS("anxious"),
    SAD("sad"),
    ANGRY("angry"),
    LONELY("lonely"),
    CONFUSED("confused"),
    GRIEVING("grieving"),
    JOYFUL("joyful"),
    SEEKING("seeking"),
    GUILTY("guilty"),
    NUMB("numb"),
    NEUTRAL("neutral");

    companion object {
        fun fromWire(value: String?): SakhaMood = entries.firstOrNull {
            it.wire.equals(value, ignoreCase = true)
        } ?: NEUTRAL
    }
}

enum class SakhaMoodTrend(val wire: String) {
    STABLE("stable"),
    RISING("rising"),
    FALLING("falling"),
    MASKED("masked");

    companion object {
        fun fromWire(value: String?): SakhaMoodTrend = entries.firstOrNull {
            it.wire.equals(value, ignoreCase = true)
        } ?: STABLE
    }
}

/**
 * Languages Sakha can speak. The wire format matches Sarvam TTS language codes
 * where applicable; STT uses BCP-47 via [sttLocale].
 */
enum class SakhaLanguage(val wire: String, val sttLocale: Locale) {
    ENGLISH("en", Locale.US),
    HINDI("hi", Locale("hi", "IN")),
    HINGLISH("hinglish", Locale("hi", "IN")),
    TAMIL("ta", Locale("ta", "IN")),
    TELUGU("te", Locale("te", "IN")),
    BENGALI("bn", Locale("bn", "IN")),
    MARATHI("mr", Locale("mr", "IN")),
    SANSKRIT("sa", Locale("hi", "IN")); // STT falls back to Hindi for Sanskrit

    companion object {
        fun fromWire(value: String?): SakhaLanguage = entries.firstOrNull {
            it.wire.equals(value, ignoreCase = true)
        } ?: ENGLISH
    }
}

// ============================================================================
// State Machine
// ============================================================================

/**
 * Sakha Voice mode state machine. A single top-level state captures
 * "where the user is in the breathing of a turn".
 */
enum class SakhaVoiceState {
    UNINITIALIZED,
    IDLE,                 // Ready to be activated
    LISTENING,            // Mic open, capturing user speech
    TRANSCRIBING,         // STT finalizing transcript
    REQUESTING,           // SSE request open, waiting for first frame
    SPEAKING,             // TTS playback in progress
    PAUSING,              // Honouring a <pause:*> marker
    INTERRUPTED,          // User barged in mid-Sakha-utterance
    ERROR,
    SHUTDOWN
}

// ============================================================================
// Configuration
// ============================================================================

/**
 * Sakha Voice configuration. Sensible defaults match the persona spec —
 * change only when you know what you are doing.
 */
data class SakhaVoiceConfig(
    /** User-facing language. Drives STT locale and TTS voice routing. */
    val language: SakhaLanguage = SakhaLanguage.ENGLISH,

    /** Backend base URL, e.g. "https://api.kiaanverse.com" — no trailing slash. */
    val backendBaseUrl: String,

    /** Optional bearer token resolver. Called once per request. */
    val authTokenProvider: suspend () -> String? = { null },

    /** Path prefix to the voice-companion routes. Defaults to MindVibe layout. */
    val voiceCompanionPathPrefix: String = "/api/voice-companion",

    /** Sarvam-style voice id for the persona body voice. */
    val sakhaVoiceId: String = "sarvam-aura",

    /** Voice id used when speaking Sanskrit verses (slower, reverent). */
    val sanskritVoiceId: String = "sarvam-meera",

    /** Default engine if backend does not route. */
    val defaultEngine: SakhaEngine = SakhaEngine.FRIEND,

    /**
     * Allow the user to interrupt Sakha mid-utterance by speaking again
     * ("barge-in"). Disabled in low-power mode.
     */
    val allowBargeIn: Boolean = true,

    /** Silence (ms) after which the mic auto-closes. */
    val silenceTimeoutMs: Long = 1800L,

    /** Hard cap on a single utterance from Sakha. Defends UX in degraded modes. */
    val maxResponseSpokenMs: Long = 45_000L,

    /** Connect / read timeout for the SSE request. */
    val requestTimeoutMs: Long = 12_000L,

    /** How many times to retry the SSE request on transient failure. */
    val maxRequestRetries: Int = 2,

    /** Run the persona guard on every chunk before TTS hand-off. */
    val enablePersonaGuard: Boolean = true,

    /**
     * If true, the manager will speak a soft template ("रहो… let us breathe
     * together for a moment") instead of staying silent on FILTER_FAIL.
     */
    val speakOnFilterFail: Boolean = true,

    /** Verbose logcat. Off in release builds. */
    val debugMode: Boolean = false,
)

// ============================================================================
// Telemetry / Metrics
// ============================================================================

/**
 * Per-turn metrics. Surface these to your observability stack but never log
 * the user's transcript.
 */
data class SakhaTurnMetrics(
    val sessionId: String?,
    val engine: SakhaEngine,
    val mood: SakhaMood,
    val moodIntensity: Int,
    val language: SakhaLanguage,
    val transcriptChars: Int,
    val responseChars: Int,
    val sttDurationMs: Long,
    val firstByteMs: Long,
    val firstAudioMs: Long,
    val totalSpokenMs: Long,
    val pauseCount: Int,
    val verseCited: String?,
    val filterFail: Boolean,
    val personaGuardTriggered: Boolean,
    val barged: Boolean,
)

// ============================================================================
// Listener
// ============================================================================

/**
 * UI-facing listener. All callbacks fire on the main thread.
 *
 * - [onStateChanged] is the canonical signal for animating the mandala.
 * - [onPartialTranscript] is best-effort; do not block on it.
 * - [onSakhaText] streams full SSE token chunks (post-pause-marker stripping)
 *   so the UI can render a transcript while audio plays.
 * - [onPause] fires every time the player honours a <pause:*> marker.
 * - [onError] errors are non-fatal unless the state ends in [SakhaVoiceState.ERROR].
 */
interface SakhaVoiceListener {
    fun onStateChanged(state: SakhaVoiceState, previousState: SakhaVoiceState) {}
    fun onPartialTranscript(text: String) {}
    fun onFinalTranscript(text: String) {}
    fun onEngineSelected(engine: SakhaEngine, mood: SakhaMood, intensity: Int) {}
    fun onSakhaText(textDelta: String, isFinal: Boolean) {}
    fun onSpokenSegment(text: String, isSanskrit: Boolean) {}
    fun onPause(durationMs: Long) {}
    fun onVerseCited(reference: String, sanskrit: String?) {}
    fun onFilterFail() {}
    fun onTurnComplete(metrics: SakhaTurnMetrics) {}
    fun onError(error: SakhaVoiceError) {}
}

// ============================================================================
// Errors
// ============================================================================

sealed class SakhaVoiceError(message: String) : Exception(message) {
    object PermissionDenied : SakhaVoiceError("RECORD_AUDIO permission not granted")
    object MicrophoneUnavailable : SakhaVoiceError("Microphone not available")
    object SpeechRecognitionUnavailable : SakhaVoiceError("Speech recognition not available")
    data class NetworkError(val cause: String) : SakhaVoiceError("Network error: $cause")
    data class HttpError(val status: Int, val detail: String) : SakhaVoiceError("HTTP $status: $detail")
    object AuthRequired : SakhaVoiceError("Authentication required")
    object QuotaExceeded : SakhaVoiceError("Daily quota exceeded")
    object FilterFail : SakhaVoiceError("Sakha could not connect retrieved verses to the prompt")
    data class TtsError(val detail: String) : SakhaVoiceError("TTS error: $detail")
    data class Unknown(val detail: String) : SakhaVoiceError(detail)

    val isRecoverable: Boolean
        get() = when (this) {
            is PermissionDenied,
            is MicrophoneUnavailable,
            is SpeechRecognitionUnavailable,
            is AuthRequired,
            is QuotaExceeded -> false
            else -> true
        }
}

// ============================================================================
// Verse recitation
// ============================================================================

/**
 * One spoken segment of a Bhagavad Gita verse recitation. The
 * [language] determines TTS routing in [SakhaTtsPlayer]:
 *
 *   - SANSKRIT             → sanskritVoiceId (slower, reverent prosody)
 *   - any other            → sakhaVoiceId   (the persona body voice)
 *
 * The [SakhaLanguage] enum already covers en / hi / hinglish / ta / te /
 * bn / mr / sa, and the TTS player resolves voice id by language without
 * further branching, so adding Kannada / Gujarati / Malayalam / Punjabi
 * later only requires an enum entry plus a TTS voice id mapping — never
 * a reader change.
 */
data class VerseSegment(
    val language: SakhaLanguage,
    val text: String,
)

/**
 * A request to recite a verse in N languages. The reader replays the
 * segments in the order supplied, so callers choose the study order:
 * Sanskrit → Hindi → English (the canonical Gita order, most users)
 * or English → Sanskrit (first-time listeners who want meaning first).
 *
 * Inserted between every two consecutive segments is a soft pause
 * (default 700ms) so the listener can absorb each language before the
 * next begins. The first and last segment have no extra leading /
 * trailing pause — the player's natural prosody closes the recitation.
 */
data class VerseRecitation(
    val chapter: Int,
    val verse: Int,
    val segments: List<VerseSegment>,
    val betweenSegmentsPauseMs: Long = 700L,
) {
    val citation: String get() = "BG $chapter.$verse"

    init {
        require(chapter in 1..18) { "BG chapter must be 1..18, was $chapter" }
        require(verse in 1..78) { "BG verse must be 1..78, was $verse" }
        require(segments.isNotEmpty()) { "VerseRecitation requires at least one segment" }
        require(betweenSegmentsPauseMs >= 0) {
            "betweenSegmentsPauseMs must be non-negative, was $betweenSegmentsPauseMs"
        }
    }
}

File: kiaanverse-mobile/apps/mobile/native/ios/Sources/Sakha/SakhaTypes.swift
/**
 * Sakha Voice — Shared Types (iOS port)
 *
 * Direct port of apps/mobile/native/android/.../SakhaTypes.kt. Enum wire
 * formats, the SakhaVoiceConfig struct, error cases, and the listener
 * protocol must stay symmetric — JS event payloads on iOS and Android
 * MUST decode identically through apps/mobile/types/sakhaVoice.ts.
 */

import Foundation

// MARK: - Engine Selection

@objc public enum SakhaEngine: Int {
    case guidance, friend, assistant, voiceGuide

    public var wire: String {
        switch self {
        case .guidance: return "GUIDANCE"
        case .friend: return "FRIEND"
        case .assistant: return "ASSISTANT"
        case .voiceGuide: return "VOICE_GUIDE"
        }
    }

    public static func fromWire(_ value: String?) -> SakhaEngine {
        switch value?.uppercased() {
        case "GUIDANCE": return .guidance
        case "ASSISTANT": return .assistant
        case "VOICE_GUIDE": return .voiceGuide
        default: return .friend
        }
    }
}

// MARK: - Mood

@objc public enum SakhaMood: Int {
    case anxious, sad, angry, lonely, confused, grieving, joyful, seeking, guilty, numb, neutral

    public var wire: String {
        switch self {
        case .anxious: return "anxious"
        case .sad: return "sad"
        case .angry: return "angry"
        case .lonely: return "lonely"
        case .confused: return "confused"
        case .grieving: return "grieving"
        case .joyful: return "joyful"
        case .seeking: return "seeking"
        case .guilty: return "guilty"
        case .numb: return "numb"
        case .neutral: return "neutral"
        }
    }

    public static func fromWire(_ value: String?) -> SakhaMood {
        guard let v = value?.lowercased() else { return .neutral }
        switch v {
        case "anxious": return .anxious
        case "sad": return .sad
        case "angry": return .angry
        case "lonely": return .lonely
        case "confused": return .confused
        case "grieving": return .grieving
        case "joyful": return .joyful
        case "seeking": return .seeking
        case "guilty": return .guilty
        case "numb": return .numb
        default: return .neutral
        }
    }
}

// MARK: - Language

@objc public enum SakhaLanguage: Int {
    case english, hindi, hinglish, tamil, telugu, bengali, marathi, sanskrit

    public var wire: String {
        switch self {
        case .english: return "en"
        case .hindi: return "hi"
        case .hinglish: return "hinglish"
        case .tamil: return "ta"
        case .telugu: return "te"
        case .bengali: return "bn"
        case .marathi: return "mr"
        case .sanskrit: return "sa"
        }
    }

    public var sttLocaleId: String {
        switch self {
        case .english: return "en-US"
        case .hindi, .hinglish: return "hi-IN"
        case .tamil: return "ta-IN"
        case .telugu: return "te-IN"
        case .bengali: return "bn-IN"
        case .marathi: return "mr-IN"
        case .sanskrit: return "hi-IN"   // STT falls back to Hindi for Sanskrit
        }
    }

    public static func fromWire(_ value: String?) -> SakhaLanguage {
        guard let v = value?.lowercased() else { return .english }
        switch v {
        case "en": return .english
        case "hi": return .hindi
        case "hinglish": return .hinglish
        case "ta": return .tamil
        case "te": return .telugu
        case "bn": return .bengali
        case "mr": return .marathi
        case "sa": return .sanskrit
        default: return .english
        }
    }
}

// MARK: - State Machine

@objc public enum SakhaVoiceStateRaw: Int {
    case uninitialized, idle, listening, transcribing, requesting, speaking, pausing, interrupted, error, shutdown

    public var wire: String {
        switch self {
        case .uninitialized: return "UNINITIALIZED"
        case .idle: return "IDLE"
        case .listening: return "LISTENING"
        case .transcribing: return "TRANSCRIBING"
        case .requesting: return "REQUESTING"
        case .speaking: return "SPEAKING"
        case .pausing: return "PAUSING"
        case .interrupted: return "INTERRUPTED"
        case .error: return "ERROR"
        case .shutdown: return "SHUTDOWN"
        }
    }
}

// MARK: - Configuration

public struct SakhaVoiceConfig {
    public var language: SakhaLanguage
    public var backendBaseUrl: String
    public var voiceCompanionPathPrefix: String
    public var sakhaVoiceId: String
    public var sanskritVoiceId: String
    public var defaultEngine: SakhaEngine
    public var allowBargeIn: Bool
    public var silenceTimeoutMs: Int64
    public var maxResponseSpokenMs: Int64
    public var requestTimeoutMs: Int64
    public var maxRequestRetries: Int
    public var enablePersonaGuard: Bool
    public var speakOnFilterFail: Bool
    public var debugMode: Bool
    public var enableWakeWord: Bool
    public var wakeWordPhrases: [String]
    public var wakeWordCooldownMs: Int64

    public init(
        language: SakhaLanguage = .english,
        backendBaseUrl: String,
        voiceCompanionPathPrefix: String = "/api/voice-companion",
        sakhaVoiceId: String = "sarvam-aura",
        sanskritVoiceId: String = "sarvam-meera",
        defaultEngine: SakhaEngine = .friend,
        allowBargeIn: Bool = true,
        silenceTimeoutMs: Int64 = 1800,
        maxResponseSpokenMs: Int64 = 45_000,
        requestTimeoutMs: Int64 = 12_000,
        maxRequestRetries: Int = 2,
        enablePersonaGuard: Bool = true,
        speakOnFilterFail: Bool = true,
        debugMode: Bool = false,
        enableWakeWord: Bool = false,
        wakeWordPhrases: [String] = [
            "hey sakha", "namaste sakha", "ok sakha", "sakha", "हे सखा", "सखा",
        ],
        wakeWordCooldownMs: Int64 = 1500
    ) {
        self.language = language
        self.backendBaseUrl = backendBaseUrl
        self.voiceCompanionPathPrefix = voiceCompanionPathPrefix
        self.sakhaVoiceId = sakhaVoiceId
        self.sanskritVoiceId = sanskritVoiceId
        self.defaultEngine = defaultEngine
        self.allowBargeIn = allowBargeIn
        self.silenceTimeoutMs = silenceTimeoutMs
        self.maxResponseSpokenMs = maxResponseSpokenMs
        self.requestTimeoutMs = requestTimeoutMs
        self.maxRequestRetries = maxRequestRetries
        self.enablePersonaGuard = enablePersonaGuard
        self.speakOnFilterFail = speakOnFilterFail
        self.debugMode = debugMode
        self.enableWakeWord = enableWakeWord
        self.wakeWordPhrases = wakeWordPhrases
        self.wakeWordCooldownMs = wakeWordCooldownMs
    }
}

// MARK: - Errors

public enum SakhaVoiceError: Error, CustomStringConvertible {
    case permissionDenied
    case microphoneUnavailable
    case speechRecognitionUnavailable
    case networkError(String)
    case httpError(Int, String)
    case authRequired
    case quotaExceeded
    case filterFail
    case ttsError(String)
    case unknown(String)

    public var code: String {
        switch self {
        case .permissionDenied: return "PermissionDenied"
        case .microphoneUnavailable: return "MicrophoneUnavailable"
        case .speechRecognitionUnavailable: return "SpeechRecognitionUnavailable"
        case .networkError: return "NetworkError"
        case .httpError: return "HttpError"
        case .authRequired: return "AuthRequired"
        case .quotaExceeded: return "QuotaExceeded"
        case .filterFail: return "FilterFail"
        case .ttsError: return "TtsError"
        case .unknown: return "Unknown"
        }
    }

    public var description: String {
        switch self {
        case .permissionDenied: return "RECORD_AUDIO permission not granted"
        case .microphoneUnavailable: return "Microphone not available"
        case .speechRecognitionUnavailable: return "Speech recognition not available"
        case .networkError(let d): return "Network error: \(d)"
        case .httpError(let s, let d): return "HTTP \(s): \(d)"
        case .authRequired: return "Authentication required"
        case .quotaExceeded: return "Daily quota exceeded"
        case .filterFail: return "Sakha could not connect retrieved verses to the prompt"
        case .ttsError(let d): return "TTS error: \(d)"
        case .unknown(let d): return d
        }
    }

    public var isRecoverable: Bool {
        switch self {
        case .permissionDenied, .microphoneUnavailable, .speechRecognitionUnavailable,
             .authRequired, .quotaExceeded:
            return false
        default:
            return true
        }
    }
}

// MARK: - Verse recitation

public struct VerseSegment {
    public let language: SakhaLanguage
    public let text: String
    public init(language: SakhaLanguage, text: String) {
        self.language = language
        self.text = text
    }
}

public struct VerseRecitation {
    public let chapter: Int
    public let verse: Int
    public let segments: [VerseSegment]
    public let betweenSegmentsPauseMs: Int64

    public var citation: String { "BG \(chapter).\(verse)" }

    /// init throws on validation failure (mirrors Kotlin's `require(…)`).
    public init(chapter: Int, verse: Int, segments: [VerseSegment], betweenSegmentsPauseMs: Int64 = 700) throws {
        guard (1...18).contains(chapter) else {
            throw SakhaVoiceError.unknown("BG chapter must be 1..18, was \(chapter)")
        }
        guard (1...78).contains(verse) else {
            throw SakhaVoiceError.unknown("BG verse must be 1..78, was \(verse)")
        }
        guard !segments.isEmpty else {
            throw SakhaVoiceError.unknown("VerseRecitation requires at least one segment")
        }
        guard betweenSegmentsPauseMs >= 0 else {
            throw SakhaVoiceError.unknown("betweenSegmentsPauseMs must be non-negative")
        }
        self.chapter = chapter
        self.verse = verse
        self.segments = segments
        self.betweenSegmentsPauseMs = betweenSegmentsPauseMs
    }
}

// MARK: - Telemetry / Metrics

public struct SakhaTurnMetrics {
    public let sessionId: String?
    public let engine: SakhaEngine
    public let mood: SakhaMood
    public let moodIntensity: Int
    public let language: SakhaLanguage
    public let transcriptChars: Int
    public let responseChars: Int
    public let sttDurationMs: Int64
    public let firstByteMs: Int64
    public let firstAudioMs: Int64
    public let totalSpokenMs: Int64
    public let pauseCount: Int
    public let verseCited: String?
    public let filterFail: Bool
    public let personaGuardTriggered: Bool
    public let barged: Bool
    public let thermalState: String?
}

// MARK: - Pause-parser events (used by parser/player)

public enum PauseEvent: Equatable {
    case speak(text: String, isSanskrit: Bool)
    case pause(durationMs: Int64)
    case filter
}

// MARK: - Listener Protocol

public protocol SakhaVoiceListener: AnyObject {
    func sakhaVoice(stateChanged state: SakhaVoiceStateRaw, previousState: SakhaVoiceStateRaw)
    func sakhaVoice(partialTranscript text: String)
    func sakhaVoice(finalTranscript text: String)
    func sakhaVoice(engineSelected engine: SakhaEngine, mood: SakhaMood, intensity: Int)
    func sakhaVoice(textDelta: String, isFinal: Bool)
    func sakhaVoice(spokenSegment text: String, isSanskrit: Bool)
    func sakhaVoice(pause durationMs: Int64)
    func sakhaVoice(verseCited reference: String, sanskrit: String?)
    func sakhaVoiceFilterFail()
    func sakhaVoice(turnComplete metrics: SakhaTurnMetrics)
    func sakhaVoice(error: SakhaVoiceError)
    func sakhaVoice(verseReadStarted citation: String)
    func sakhaVoice(verseSegmentRead citation: String, language: SakhaLanguage)
    func sakhaVoice(verseReadComplete citation: String)
    func sakhaVoice(wakeWord phrase: String)
}

/// Default no-op implementations so adopters only override what they need.
public extension SakhaVoiceListener {
    func sakhaVoice(stateChanged state: SakhaVoiceStateRaw, previousState: SakhaVoiceStateRaw) {}
    func sakhaVoice(partialTranscript text: String) {}
    func sakhaVoice(finalTranscript text: String) {}
    func sakhaVoice(engineSelected engine: SakhaEngine, mood: SakhaMood, intensity: Int) {}
    func sakhaVoice(textDelta: String, isFinal: Bool) {}
    func sakhaVoice(spokenSegment text: String, isSanskrit: Bool) {}
    func sakhaVoice(pause durationMs: Int64) {}
    func sakhaVoice(verseCited reference: String, sanskrit: String?) {}
    func sakhaVoiceFilterFail() {}
    func sakhaVoice(turnComplete metrics: SakhaTurnMetrics) {}
    func sakhaVoice(error: SakhaVoiceError) {}
    func sakhaVoice(verseReadStarted citation: String) {}
    func sakhaVoice(verseSegmentRead citation: String, language: SakhaLanguage) {}
    func sakhaVoice(verseReadComplete citation: String) {}
    func sakhaVoice(wakeWord phrase: String) {}
}
